hello-robot/stretch_web_teleop
# Signaling

# Description

This PR is the second of two PRs (the first is https://github.com/hello-robot/stretch_web_teleop/pull/115) that will make web teleop deployable on the internet. It introduces "signalers", which manage sending signals to establish a WebRTC connection between a robot and an operator anywhere in the world. Web teleop already had local signaling embedded in `webrtcconnections.tsx` + a signaling server in `server.js`, which means that if you could do signaling if you could access the robot (often through ngrok tunneling or tailscale). However, since it's not common to expose tunnels/tailscale to the internet, a new way of signaling was needed. In this PR, the `BaseSignaling` interface is introduced and two implementations are added: `LocalSignaling` and `FirebaseSignaling`. Whereas the former still uses `server.js`, the latter uses Firebase's Realtime Database to transfer signals. This PR builds on the previous PR's authentication to ensure security. Lastly, we can support alternate signaling servers in the future by adding new implementations of `BaseSignaling` to the `shared.signaling` module. 

# Testing procedure

We will be testing this PR and the [Login & Homepage PR](https://github.com/hello-robot/stretch_web_teleop/pull/115) together. Check out the "feature/signaling" branch. Go through the testing instructions for the other PR first, then go through these:

### Running the interface locally

Launch the interface

1. Go to `https://localhost/operator` in a browser and you should see the web interface.
    - ![image](https://github.com/user-attachments/assets/7ee88cf7-d02e-4683-ad55-dbaef1ea14e8)
2. Open the homepage in another tab. The robot will be occupied.
    - ![image](https://github.com/user-attachments/assets/e2c2cef8-ac58-4ab9-bb81-bca2ad34567a)
3. Close the operator panel. The robot will become available on the homepage.
4. Stop the robot browser, and the robot will go offline on the homepage.

### Running the interface online

You will need to send me your email address and robot's serial number before you can test this section. Additionally, you will need to ask me for an `.env` file to put in the root of your repo, change "localstorage" to "firebase" in line 4 of start_web_server_and_robot_browser.sh, and then launch the interface normally.

1. Visit https://web.hello-robot.com/ and sign in. You should see your online and available.
    - ![image](https://github.com/user-attachments/assets/359de7a3-480d-49be-b698-8a92948970df)
2. Press the "Call Robot" button and you should see the web interface.
    - ![image](https://github.com/user-attachments/assets/f43c0217-a896-4669-9f71-a647af16ad2b)
    - You can bookmark the URL to call the robot without going through the homepage
    - Open the homepage in another tab. The robot will be occupied. Close the operator panel, and the robot will become available.
3. Visit `https://web.hello-robot.com/operator` and the interface will not load. In the console, you'll see that the interface was unable to determine which robot to call because the URL was missing the `robot` parameter.
    - ![image](https://github.com/user-attachments/assets/9d8da95e-26c6-4662-bebb-bd8251afd685)
5. Visit `https://web.hello-robot.com/operator/?robot=stretch-se3-3005` to try calling a robot that doesn't belong to you. The interface will not load.
    - ![image](https://github.com/user-attachments/assets/7899e6b1-5b71-4534-bf0a-a534379270c1)
6. Stop the interface on the robot and the robot may go offline on the homepage. There is a known bug that the robot may continue to appear online and callable on the homepage, despite the robot browser exiting. This happens because Chrome/other browsers don't give `window.onbeforeunload` enough time to complete entirely. In the future, we'll use heartbeats to fix this and make connectivity more robust in general.

# Before opening a pull request

From the top-level of this repository, run:


# To merge
